# Exploring Ebay Car Sales Data
In this project, we will clean and analyze data scraped from a classifields section of the German eBay website that has not yet been cleaned. The dataset includes information about used car listings. You can find the dataset used and its documentation [here](https://data.world/data-society/used-cars-data).

In [179]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [180]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [181]:
print(autos.info(), '\n')
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Observations

There seems to be null values in the vehicleType, gearbox, model, fuelType, and notRepairedDamage columns, according to the non-null counts in autos.info(). 

The columns price and odometer should be converted from object to numeric dtypes so statistics can be applied on the values. Similarly, the date-type columns dateCrawled, yearOfRegistration, monthOfRegistration, dateCreated, and lastSeen should be converted into datetime objects so mathematical operations can be run on the numeric values of the dates and times.

The notRepairedDamage column might benefit from being converted into a boolean or 0/1 value column as it is a conditional attribute of a car. The unique values in this column are 'nein', NaN, and 'ja'.

The column titles are in camelCase- converting the columns into snake_case would improve readability.

In [182]:
print(autos['notRepairedDamage'].unique())

['nein' nan 'ja']


In [183]:
print(autos.columns)

new_cols = []
for col in autos.columns:
    if col == 'yearOfRegistration':
        temp = 'registration_year'
    elif col == 'monthOfRegistration':
        temp = 'registration_month'
    elif col == 'notRepairedDamage':
        temp = 'unrepaired_damage'
    elif col == 'dateCreated':
        temp = 'ad_created'
    else: # turn column label from camelCase to snake_case
        temp = ''
        for char in col:
            if (ord(char) >= 65) and (ord(char) <= 90): # if char is uppercase
                if (ord(temp[-1]) >= 65) and (ord(temp[-1]) <= 90): # if prev char is uppercase
                    temp += char
                else:
                    temp += '_{}'.format(char)
            else:
                temp += char
        temp = temp.lower()
    new_cols.append(temp)

autos.columns = new_cols
autos.head

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


<bound method NDFrame.head of               date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V  

In the cell above, we converted column labels from camelCase to snake_case and modified the wording of some columns to improve readability and to make it easier to reference the column labels when coding.

In [184]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [185]:
print(autos['seller'].value_counts(), '\n')
print(autos['offer_type'].value_counts(), '\n')
print(autos['abtest'].value_counts(), '\n')
print(autos['gearbox'].value_counts(), '\n')
print(autos['unrepaired_damage'].value_counts(), '\n')

print(autos['power_ps'].value_counts())

print(autos['price'].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64 

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64 

test       25756
control    24244
Name: abtest, dtype: int64 

manuell      36993
automatik    10327
Name: gearbox, dtype: int64 

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64 

0        5500
75       3171
60       2195
150      2046
140      1884
         ... 
650         1
490         1
362         1
153         1
16312       1
Name: power_ps, Length: 448, dtype: int64
$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$21,049       1
$459          1
$39,990       1
$34,940       1
$6,749        1
Name: price, Length: 2357, dtype: int64


### Observations
First thing we can notice are 5 text columns with only 2 unique values. Upon closer look, 2 of the columns have disproportionally unique values:
- The 'seller' column is comprised of 49999 'private' and 1 'gewerbilch' values
- The 'offer_type' column is comprised of 49999 'Angebot' and 1 'Gesuch' values

We want to omit these two text columns that have all or almost all of the same values as they do not contain any useful information for analysis.

The power_ps column has a min value of 0, which is a weird value for the horsepower of a car. Upon closer look, we see that a horsepower of 0 is actually the most common value for this column with 5500 entries- this is alarming. Perhaps this dataset includes cars without engines and/or cars with broken engines.

The most frequent value in the price column is $0- it's odd to list a car for free on eBay. Perhaps they are totalled cars being given away for parts, which would explain some of the 0 horsepower values in the power_ps column.

Besides these observations, there are several columns whose values should be converted into different datatypes so we can perform useful statistics on them.

In [186]:
print(autos['price'].unique())
print(autos['odometer'].unique())

try:
    autos['price'] = (autos['price']
                      .str.replace('$', '')
                      .str.replace(',', '')
                      .astype(int)
                     )
    autos['odometer'] = (autos['odometer']
                         .str.replace(',', '')
                         .str.replace('km', '')
                         .astype(int)
                        )
except:
    print('Conversion already done')

autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

print(autos['price'].unique())
print(autos['odometer_km'].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']
[ 5000  8500  8990 ...   385 22200 16995]
[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]


In [187]:
print(autos['price'].min(), '\t', autos['price'].max(),'\n')
print(autos['price'].describe())
autos['price'].value_counts().sort_index(ascending=False).head(15)

0 	 99999999 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [188]:
num_price_zero = autos[autos['price'] == 0].shape[0]
total_listings = autos.shape[0]
percent_price_zero = (num_price_zero / autos.shape[0]) * 100

# omit listings with a price of 0 or above 350,000
print(percent_price_zero)
autos = autos[autos['price'].between(1,350000)]
autos['price'].value_counts().sort_index()

2.842


1         156
2           3
3           1
5           2
8           1
         ... 
265000      1
295000      1
299000      1
345000      1
350000      1
Name: price, Length: 2346, dtype: int64

Exploring the value counts of the price column, we find that 2.8\% of the listings have a price of 0 and unique price values climb steadily from 1 to 350,000- from there, the prices jump to unrealistic numbers at 999,990+; some of these prices seem to have been arbitrarily chosen as they are funny numbers like 999,999, 1,234,566,and 111,111,111.

We'll keep listings with prices in the range 1-350,000. In the context of eBay, listings can be put up for auction starting at 1, though not 0. Listings with prices of only 0 makeup only 2.8\% of the data, so it won't affect our analysis too much to omit them; listings with unrealistic prices over 350,000 makeup even less of the data. 

In [189]:
print(autos['odometer_km'].min(), '\t', autos['odometer_km'].max(),'\n')
print(autos['odometer_km'].describe())
autos['odometer_km'].value_counts().sort_index()

print((autos[autos['odometer_km'] == 150000].shape[0] / total_listings) * 100)

5000 	 150000 

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
62.827999999999996


Odometer values range from 5000-150,000 which are reasonable, however, we must keep in mind that some of these values are ambiguous as the values were originally of the form n km+ and were categorically chosen instead of defining the exact mileage.

The odometer_km values are skewed as 62.8% of the listings have mileages of 150,000+. 

As far as we can tell, there are no outliers or invalid values that should be omitted.

In [190]:
pd.set_option('display.max_rows', None)

print(autos['ad_created']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000082
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001483
2016-03-05    0.022897
2016-03-06 

In [191]:
print(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


In [192]:
print(autos['last_seen']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)
pd.reset_option('display.max_rows')

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The earliest ad created in this dataset was on 2015-06-11 and the latest ad on 2016-04-06. Most ads seem to have been created from 2016-03-07 to 2016-04-04. There are several dates that no ads were created on outside of this range, but there were ads created on every day of March 2016 and April 2016 up to April 7th, the last day data was crawled.

The data was crawled almost a year later (2016-03-05) after the first ad was created in the dataset (2015-06-11). But other than that the data seems to have been consistently crawled from 2016-03-05 to 2016-04-07. The data crawled is mostly evenly distributed likely due to rate limiting and processing limits/rates.

The last_seen data starts and ends in the same range as when the data was crawled, of course. The normalized value counts don't give us much information as when a particular ad was created and last seen is ambiguous when looking at just the last_seen dates. The counts for the dates seems to skew towards the end of when the data was crawled, which makes sense as listings take time to be sold and will likely be sold later than sooner.


In [193]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [197]:
np.sort(autos['registration_year'].unique())

array([1000, 1001, 1111, 1800, 1910, 1927, 1929, 1931, 1934, 1937, 1938,
       1939, 1941, 1943, 1948, 1950, 1951, 1952, 1953, 1954, 1955, 1956,
       1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967,
       1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978,
       1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989,
       1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2800, 4100, 4500,
       4800, 5000, 5911, 6200, 8888, 9000, 9999])

Invalid values for car registration years exist such as 1000, 1001, 1111 before cars were created and wacky numbers like 8888, 9000, 9999. Furthermore, values greater than the latest year the ads were created (2016) don't make sense either as the registration year must be in the present, not in the future. 

In [215]:
bad_reg_year_counts = (autos.loc[~autos['registration_year'].between(1886, 2016), 'registration_year']
 .value_counts(dropna=False)
 .sort_index()
)

print((bad_reg_year_counts.sum() / total_listings) * 100)

bad_reg_year_counts

3.768


1000       1
1001       1
1111       1
1800       2
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

In [216]:
autos = autos[autos['registration_year'].between(1886, 2016)]

In [223]:
pd.set_option('display.max_rows', None)
print(autos['registration_year']
      .value_counts(normalize=True)
      .sort_index()
     )
pd.reset_option('display.max_rows')

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
1969    0.000407
1970    0.000814
1971    0.000557
1972    0.000707
1973    0.000493
1974    0.000514
1975    0.000386
1976    0.000450
1977    0.000471
1978    0.000943
1979    0.000728
1980    0.001821
1981    0.000600
1982    0.000878
1983    0.001093
1984    0.001093
1985    0.002035
1986    0.001542
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007262
1992    0.007926
1993    0.009104
1994    0.013474
1995    0.026285
1996    0.029412
1997    0.0417

Above, we decided that valid registration years are anytime from 1886, the year the first automobile (Benz motor car) was patented, to 2016, the latest year the data was crawled. We ommitted values outside this range.

The  of registration years grow as they approach the date the year the data was distributioncrawled, peaking at 2005 when 6.3% of cars were registered and starts to decline approaching 2015 and 2016. This makes sense as this data is from the German eBay used cars market- used cars will typically be of older, but not vintage year models. Furthermore, it's less likely cars bought in the current or last year would be resold so soon.

In the German eBay used car market, most used cars listed were registered in 2005.

In [252]:
avg_price_per_top_brands = {}
for brand in autos['brand'].value_counts(normalize=True)[:10].index:
    avg_price = autos.loc[autos['brand'] == brand, 'price'].mean()
    avg_price_per_top_brands[brand] = avg_price
avg_price_per_top_brands

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287,
 'renault': 2474.8646069968195,
 'peugeot': 3094.0172290021537,
 'fiat': 2813.748538011696,
 'seat': 4397.230949589683}

We have found that German brands make up 4 out of the 5 most frequent brands in the German eBay used car market. Above, we aggregated the 10 most frequent brands and their average prices. Focusing on the popular German brands Volkswagen, BMW, and Audi, we find that the luxury BMW and Audi vehicles are thousands more on average at 8.3k-9.3k than Volkswagen at 5.4k. In fact, of those three, their popularity in the market is negatively correlated with their average price with Volkswagen being the most popular and the cheapest on average, and Audi being the least popular (of the three) and the most expensive on average. 

Volkswagen may very well be the most popular because of its more accessible, middle-ground price point compared to less-popular but cheaper Opel, Renauly, and Peugeot brands, whose vehicles are <3.1k on average, and the higher-coveted Audi and BMW at >8.3k on average.

Besides the highly dominant German brands are the Ford, Fiat and Seat brands that are less popular and less expensive on average than Volkswagen, BMW, and Audi.

In [253]:
avg_mileage_per_top_brands = {}
for brand in autos['brand'].value_counts()[:10].index:
    avg_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    avg_mileage_per_top_brands[brand] = avg_mileage
avg_mileage_per_top_brands

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558,
 'seat': 121131.30128956624}

In [258]:
bmp_series = pd.Series(avg_price_per_top_brands)
bmm_series = pd.Series(avg_mileage_per_top_brands)
brand_agg = pd.DataFrame(bmp_series, columns=['mean_price'])
brand_agg['mean_mileage'] = bmm_series
brand_agg

mean_price   mean_mileage
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872
renault        2474.864607  128071.331213
peugeot        3094.017229  127153.625269
fiat           2813.748538  117121.971596
seat           4397.230950  121131.301290

Just as there are so many attributes/columns per used car listing, there are other potential factors that may influence the average price of a given brand of car. We'll extend our aggregation to the average mileage per brand to see if we can find a correlation to the average price.

For our 10 most popular brands, the average mileage per car brand are actually mostly the same with the highest average being 132,573 miles for BMW and the lowest being 117,122 miles for Fiat- a negligible few ten thousand miles in mileage. Therefore, we can say that the mileage has little correlation with the average price per car brand. The quality, luxuriousness, and reputation of each brand is likely the leading factor to the car's average price, in